extensions/replication/tasks/MultipleBackendTask.js

```diff
@@ -736,93 +784,117 @@ class MultipleBackendTask extends ReplicateObject {
         log.debug('getting object data', { entry: sourceEntry.getLogInfo() });
         const doneOnce = jsutil.once(done);
         const size = sourceEntry.getContentLength();
-        let sourceReq = null;
-        let incomingMsg = null;
-        let aborted = false;
+        
+        const abortController = new AbortController();
+        let sourceStreamAborted = false;
+        
+        const performPutObject = incomingMsg => {
+            log.debug('putting object', { entry: sourceEntry.getLogInfo() });
+            
+            // Set up stream error handler if we have a stream
+            if (incomingMsg) {
+                incomingMsg.on('error', err => {
+                    if (!sourceStreamAborted) {
+                        sourceStreamAborted = true;
+                        abortController.abort();
+                        // eslint-disable-next-line no-param-reassign
+                        err.origin = 'source';
+                        if (err.$metadata?.httpStatusCode === 404) {
+                            log.error('the source object was not found', {
+                                method: 'MultipleBackendTask._getAndPutObjectOnce',
+                                entry: sourceEntry.getLogInfo(),
+                                origin: 'source',
+                                peer: this.sourceConfig.s3,
+                                error: err.message,
+                                httpStatus: err.$metadata?.httpStatusCode,
+                            });
+                            return doneOnce(errors.ObjNotFound);
+                        }
+                        log.error('an error occurred when streaming data from S3', {
+                            entry: sourceEntry.getLogInfo(),
+                            method: 'MultipleBackendTask._getAndPutObjectOnce',
+                            origin: 'source',
+                            peer: this.sourceConfig.s3,
+                            error: err.message,
+                        });
+                        return doneOnce(err);
+                    }
+                    return undefined;
+                });
+            }
+            
+            const putDone = err => {
+                if (err && incomingMsg && !sourceStreamAborted) {
+                    // Abort the source stream on PUT error
+                    sourceStreamAborted = true;
+                    abortController.abort();
+                    if (incomingMsg.destroy) {
+                        incomingMsg.destroy();
+                    }
+                }
+                return doneOnce(err);
+            };
+            
```

Comment:
similarly, this function looks similar, but diff (and new function) is now much harder to read
because of the reversed logic...

maybe using asyc code wrapped in callbackify would allow keeping logical order...

Reply:
or we could use `async.waterfall`, which allows mixing continuation-cb methods and async methods:

```suggestion
async.waterfall([
    async () => {
        if (size == 0) {
            return undefined;
        }

        const command = ...
        const data = await this.backbeatSource.send(command, { abortSignal: abortController, signal })
        return incomingMsg.Body;
    },
    (incomingMsg, cb) => {
        if (sourceEntry.getReplicationIsNFS()) {
            return this._checkObjectState(sourceEntry, log, err => {
                if (err) {
                    return cb(err);
                }
                return cb(null, incomingMsg)
            }
        }
        return cb(incomingMsg)
    },
    (incomingMsg, cb) => this._sendMultipleBackendPutObject(sourceEntry, size, incomingMsg, log, cb);
], putDone);
```

(could be used for other cases above, to keep the flow)